### ISSUE TITLE
Instructions for creating an API key and OAuth credentials

### ISSUE THREAD

**Issue by tl-pierre** [anxdpanic/plugin.video.youtube#439]
Hi,
On the [Personal API Keys wiki page ](https://github.com/anxdpanic/plugin.video.youtube/wiki/Personal-API-Keys), the tutorial video for generating API keys is down : "This video has been removed for violating YouTube's Terms of Service"
Video URL : [https://www.youtube.com/watch?v=4iFaY8IpTeI](https://www.youtube.com/watch?v=4iFaY8IpTeI)
Currently, there is no other explanation than this video tutorial in the wiki unfortunately.

Thank you for this amazing Kodi plugin :heart_eyes:

**Comment by slax57** [comment]
Here are my 2 cents for this issue.

I found this great video explaining the process, unfortunately it's in French 😁 .
I figured it might be useful to some, though, so here it is:

\>>> https://www.youtube.com/watch?v=kKMFH5SwyNw <<<

As a complement, here is a small written tutorial with the steps.

## API Key Creation

1. Go to https://console.developers.google.com/ and create a new project

![Capture](https://user-images.githubusercontent.com/14542336/232154445-b3b2d55c-0d89-47cb-8729-c14880205c2a.PNG)

2. Click on `Activate APIs and Services`

![Capture2](https://user-images.githubusercontent.com/14542336/232154646-ee1a63a6-0adf-4e6d-a6fe-58c538dbb864.PNG)

3. Activate `Youtube Data API v3`

![Capture3](https://user-images.githubusercontent.com/14542336/232154765-b9a3201a-bfa3-4eee-a3d3-70dfb2019cd8.PNG)

4. Go to the `Credentials` section and click `Create new credentials`

5. Create a new credential of type `API Key`. Choose a name, and leave the restriction choices to `No restriction`

![Capture4](https://user-images.githubusercontent.com/14542336/232155153-d41d9bc2-adb5-4d99-92dd-e3bf8f6a414a.PNG)

6. Create a new credential of type `OAuth client Id`

![Capture5](https://user-images.githubusercontent.com/14542336/232155359-c6d012a5-70be-4cce-8f64-d737b9bf4555.PNG)

7. Choose `External` user type

![Capture6](https://user-images.githubusercontent.com/14542336/232155513-b07c8c42-1f07-4177-9ca7-5785bb164069.PNG)

8. Choose any name for the App. Select your email address in the list. Leave all the other options to the default.

![Capture7](https://user-images.githubusercontent.com/14542336/232155763-a7d5b357-a357-4c11-a7d8-c3254bf22be3.PNG)

9. Create a new credential of type `OAuth client Id`, _again_

![Capture5](https://user-images.githubusercontent.com/14542336/232155359-c6d012a5-70be-4cce-8f64-d737b9bf4555.PNG)

10. Choose `TV input devices and limited` as application type, and leave the client name to default

![Capture8](https://user-images.githubusercontent.com/14542336/232156215-4388bf50-b137-4e75-a3d1-08f262f83894.PNG)

11. Go to the `OAuth consent screen` section, and click the `Publish` button to publish you app

![172390662-ba4c2e93-6b19-4353-a485-4cc7fd6ab8f6](https://user-images.githubusercontent.com/14542336/232158749-91935403-88b3-4704-8513-b5a20e963663.png)

Alternatively you can add your Account as a Test User, thereby allowing access. By default nobody (not even the account owning the project) has access.

![198346405-9c4328a9-7b35-450e-87cd-eb5a09eeefa9](https://user-images.githubusercontent.com/14542336/232158853-9e815174-187a-4fd8-affe-50e56f885772.png)

Kudos to https://github.com/anxdpanic/plugin.video.youtube/issues/312 for this method

12. Back to the `Credentials` page, you should now have access to all 3 required keys

![Capture9](https://user-images.githubusercontent.com/14542336/232157106-91118abf-1ec7-48ad-923f-38b9cb9a2a5f.PNG)

13. To get the `Client secret`, you need to edit the OAuth credential. This should open a popper window including the secret.

![Capture10](https://user-images.githubusercontent.com/14542336/232157366-c27be873-d180-433d-b5d8-d08688bbcc22.PNG)

## Configure the Add-on

Follow the instructions in the wiki section: https://github.com/anxdpanic/plugin.video.youtube/wiki/Personal-API-Keys#configure-the-add-on

## Sign In

During the add-on Sign in process you'll be asked to `Connect a device` twice.

The second time you `Connect a device`, you will be prompted with:

```
This app isn't verified
This app hasn't been verified by Google yet. Only proceed if you know and trust the developer.
```

From there clicking `Advanced` and `Go to <your project name> (unsafe)` allows you to complete the process.

![Capture11](https://user-images.githubusercontent.com/14542336/232159697-b5274507-8f18-450e-a306-9d59f8b1ba79.PNG)

This should be it! 🎉 🎉 🎉 

## Additional notes

I have noticed that I had to run the whole `Sign In` process a second time after I exited Kodi and started it back again, i.e. perform the `Connect a device` action 4 times total.

(this may be related to https://github.com/anxdpanic/plugin.video.youtube/issues/388)

But the good new is, after all this, Kodi now finally remembers my account each time I start it up 🎉 🎉 🎉 

I hope this little guide will be helpful to some of you, and why not, could help updating the wiki 😇

**Comment by thecyberd3m0n** [comment]
just reupload it to other service. Vimeo, Odysee or other

**Comment by Otiel** [comment]
Thanks for the detailed guide @slax57! ❤️ 

I also found [this guide](https://raspberrytips.com/youtube-api-keys-kodi/) helpful for those looking for english screenshots.

**Comment by DavidFW1960** [comment]
Is there any fix to make this work if you have 2 factor authentication enabled on the account?

**Comment by MoojMidge** [comment]
What doesn't work when you have 2FA enabled?

You need to authenticate account access via a browser to grant Kodi the OAuth access token, and it is only when you login in the browser that you would need to provide the details for the 2FA.

Unless something has changed?

**Comment by DavidFW1960** [comment]
> What doesn't work when you have 2FA enabled?
> 
> You need to authenticate account access via a browser to grant Kodi the OAuth access token, and it is only when you login in the browser that you would need to provide the details for the 2FA.
> 
> Unless something has changed?

I am getting a 400 invalid_token error

**Comment by MoojMidge** [comment]
The error is fairly self explanatory - it means that the token you are using is not valid. 

This usually will happen when you created the wrong type of token, or are using an old token that is no longer valid that causes the revoke endpoint to return this error when trying to refresh the access token.

This is just a guess because you are not really providing any useful details to diagnose what the problem could be.

You should create a new project/API key with the proper scope, and create new credentials. Enter the new API key details into the addon settings and try signing in again.

If you are still having the same problem create a new issue with full details and a debug log.

**Comment by DavidFW1960** [comment]
I reset every page of the addon settings to default and still same error. My token is years old. I will try and create new credentials and try again

**Comment by MoojMidge** [comment]
> My token is years old.

Yeah that is the likely cause, they can be revoked due to lack of use, change in scope, on Google's whim, etc.

Just to be sure, make a new API key as well, not just new credentials for an old key.

**Comment by DavidFW1960** [comment]
ok I redid the auth and had to log in twice. Seemingly successful. But I get:
```
error <general>: [plugin.video.youtube] Error reason: |quotaExceeded| with message: |The request cannot be completed because you have exceeded your quota.|```
Hard to see how that's possible given I only just got it auth'ed
I also get list for my Subscriptions.
Here is my log
https://1drv.ms/t/s!An_0ZtrQyrRIhMI_aOKx5ldky-PyGA?e=axfKnn

**Comment by MoojMidge** [comment]
The log is a little strange. You only seem to have logged out and logged back in 7 minutes before closing Kodi, but the quota exceeded errors were occuring well before that.

It almost sounds like your API key has been disabled, but you didn't create a new API key, just a new valid OAuth credential for the existing disabled API key, which is why you are no longer getting the invalid_token_error but the quotaExceeded error instead.

It is not uncommon for this to occur which is why I recommended creating a new API key, not just a new OAuth credential.

> ok I redid the auth and had to log in twice

If you are absolutely sure you created a new API key before this, then one thing to check is to go to https://console.cloud.google.com/apis/api/youtube.googleapis.com/quotas and check what your quota allocaton and usage actually is. You should see something like the table below:

![image](https://github.com/anxdpanic/plugin.video.youtube/assets/56883549/f44b06c2-c4b9-4299-9ff0-7714d22d3a2d)

In some instances a newly created API key may have close to 100% of the daily quota allowance already used, which resets after a period of time (at most one day). Alternatively if you have a disabled API key, you will see that your quote limit is set to zero, and your only option is to create a new API key.

Edit: Also update the addon from here: https://github.com/anxdpanic/plugin.video.youtube/releases/latest
v7.0.1 likely won't work even if you do resolve the API key and OAuth login issue. Newer versions may not have those issues at all.

**Comment by DavidFW1960** [comment]
Quota is 0. I did create a new API. I updated yesterday to the new version as well. I'm going to try making a new project etc again....

**Comment by MoojMidge** [comment]
Hmm that sucks. It happens sometimes for seemingly no reason, and worst of all there is no advance warning either.

> I'm going to try making a new project etc again....

Yeah unfortunately creating a new API key is your only option, and creating a new project is probably a good idea too.

**Comment by DavidFW1960** [comment]
I DID create a new API key yesterday as you suggested and today I did the new project and I have a 10000 quota and it's working. Finally. Big thanks for this addon and all your help.

**Comment by MoojMidge** [comment]
> I DID create a new API key yesterday

I'll admit I was suspicious because the process is not straightforward and it is easy to not know the reason/significance of each step, but I do believe you (at least now 🙂).

Why it happens, even for a new API key, is something only Google can answer, but it's good you got it working despite the hassle.

**Comment by DavidFW1960** [comment]
Yeah i had it working years ago before I switched to using 2FA and it hasn't worked since. It was on the backburner but made it to the front now….

**Comment by cg00001** [comment]
![1](https://github.com/user-attachments/assets/ce1bbd98-1bec-42ea-b65f-415c1d0e6f9d)

This is what I did -and worked- when Youtube had a quota frenzy on me last year:
https://console.cloud.google.com/apis/dashboard
1. Check traffic and quotas clicking on Enabled APIs & services. If your API was left on a build or Youtube is playing dirty is irrelevant. You will change your API and Youtube Kodi addon will work -at least for a while.
2. Click on Credentials.
3. Click on CREATE CREDENTIALS.
4. Click on API key.
After it's created, click on it, it's suggested to limit it to work only with YouTube Data API v3.
Make sure to save it, it's the one that will replace the old one in your api_keys.json which is in your /.kodi/userdata/addon_data/plugin.video.youtube/ folder.
You can save api_keys.json somewhere and once you install/update a build or plain Kodi, just install Youtube addon, and replace it with yours to easily sign in.

**Comment by damnms** [comment]
Would be nice if the tutorial could be updated. The pictures show very different stuff that is shown on https://console.cloud.google.com/ - it starts with "create a new project" - there is no such button anymore. At least not on the dashboard page when i visit https://console.cloud.google.com/

**Comment by MoojMidge** [comment]
You see all the other issues linked to this one? 

For the most part they are issues that a whole bunch of people have raised, without bothering to search and read, and who have also never bothered to contribute back by documenting what they did, as they completed these steps for themselves, so that the instructions and screenshots could remain up-to-date for others.

This addon has a large number of users. Some contribute by testing and reporting issues or otherwise providing logs/feedback, and some have provided updated instructions here and on the Kodi forums. My thanks goes out to all of them, especially to @slax57 whose written instructions I continue to link to because they are the most detailed and clear, despite being a little out of date.

Unfortunately the majority of users contribute nothing while often expecting a lot. The lack of contributions doesn't bother me too much, no one is obligated to do anything, but it also means I am not particularly motivated to do things that don't interest me, and this particular task does not interest me.

If anyone is sufficiently motivated, the following would be nice:

- an updated instructional video without advertising or discussion of other addons or services.
- an updated step by step written guide, in any language, similar to https://github.com/anxdpanic/plugin.video.youtube/issues/439#issuecomment-1509319017, with updated screenshots.
- an updated step by step written guide, in any language, outlining the steps required to enter the API keys and OAuth details and the sign in using the addon.
- deep links to the console pages that perform the required actions necessary for each step.

**Comment by damnms** [comment]
I would do that if i got it working. I am not publishing something that does not work on my side, because that would make things even worse when it does not work. I really appreciate your work and if i can do something, i'd love to help. Unfortunately, my skills are rather... limited.
edit: got it working, thanks, i had to login which i never had to do before (thought api key/share is exactly for that). I will create a small tutorial which might help some people.

**Comment by rrawstron** [comment]
@MoojMidge I authored the click by click instructions for API key creation and YouTube add-on authentication here
https://forum.kodi.tv/showthread.php?tid=356934&pid=3224383#pid3224383

You're welcome to copy this content to https://github.com/anxdpanic/plugin.video.youtube/wiki/Personal-API-Keys

**Comment by MoojMidge** [comment]
Thank you @rrawstron. I saw the post on the forums but haven't had a chance to do anything with it.

As I mentioned to @damnms, I have no real motivation to do so, but seeing as how a number of people here and on the forums have continued to provide assistance, the least I can do is help make this easier by updating things in a central location.

Just quite busy irl at the moment.

**Comment by damnms** [comment]
@rrawstron i created a little updated "guide" with pictures (because in other guides they were either not in english or outdated), if you can use any of the screenshots i created, feel free to use them: https://github.com/anxdpanic/plugin.video.youtube/issues/1095. I am wondering if it would be possible to somehow automate all this. I am sure its a ton of work with google api's but that would be awesome

**Comment by rrawstron** [comment]
@MoojMidge Ok cool you had seen the Kodi forum post (which is really all I wanted to check). I totally understand about being busy and having other priorities so I'll leave this with you to deal with in your own time.

@damnms Sorry I didn't mean to duplicate your effort.
That is an interesting question about can the GCP config be automated (to enable an easier setup).
I'm not a GCP expert by any means. A quick Google shows these options for project exporting/importing options:
- https://cloud.google.com/config-connector/docs/how-to/import-export/export
- https://cloud.google.com/docs/terraform/resource-management/export
- https://github.com/GoogleCloudPlatform/terraformer

Though each of these requires setting up an environment to enable the export/import.
Given the broad range of users who want to set this up I'd personally lean towards just providing basic GCP console click by click instructions as the lowest common denominator (which will become out of date as soon as they typed up ;-D).

However now I'm curious. If I have time I might look at these options for myself...

**Comment by wnto** [comment]
Hi, I spent a few hours yesterday trying to login to youtube through after creating and adding the api keys. I have a project with external audience, with my google user added as a test user, api key with no restrictions and, no matter what I try, after I do the second round of device authorization (through entering the codes displayed on kodi), google blocks me from going any further (403), because my app has not been reviewed. And this is trying to login with an authorized test user. the google message (attached) mentions that this is due to my google account having a 2nd factor authentication. 
I could test it with an account without 2FA, but that wouldn't solve things, as the reason I went through all these hoops was to get my user history and preferences running on my TV. Also, I just got extremely tired doing all that. 

<img width="576" height="1280" alt="Image" src="https://github.com/user-attachments/assets/1baa5391-0343-4803-8071-89e41b24a467" />

My whole idea was to buy a new tv, never connect it to the internet, add the services that I mostly use (youtube, netflix, prime) to a linux machine I control and, slowly, migrating from data hungry streaming services to more private solutions and local libraries. But I cannot do the hard switch, I would get quickly outvoted at home. 
I understand that companies might be making these integrations intentionally difficult, but, if so, they are succeeding in killing 
off the viability of open and free alternatives for the majority of people.

**Comment by MoojMidge** [comment]
1. > with my google user added as a test user

2. > And this is trying to login with an authorized test user

3. > the google message (attached) mentions that this is due to my google account having a 2nd factor authentication.

There is no message attached. The screenshot indicates that either statements from `#1` or `#2` are not correct.

Only thing I can advise you to do is to double check the accounts that you have added as test users. It will have nothing to do with 2FA.

**Comment by wnto** [comment]
thank you, I will try once more. In any case, I didn't mean to have sentences 1 and 2 be different. I added the user I used to create the APIs as an authorized test user. And checked this more than once. Nonetheless, I got the message shown in the screenshot, which only made me even more confused. 
I'll try it again eventually (travel plans), and let you know if it works.

**Comment by tired9494** [comment]
<img width="646" height="364" alt="Image" src="https://github.com/user-attachments/assets/a170f30f-491f-48d6-8401-4e2ccb263ed3" />

Less secure app access cannot be enabled for accounts with 2FA, and there is nowhere to enter an app password ( https://support.google.com/accounts/answer/185833?hl=en-GB&sjid=11292038056175908941-EU )

**Comment by MoojMidge** [comment]
Less secure app access is not required.

**Comment by tired9494** [comment]
That part should be removed from the wiki then. idk why else it's not working for me, I'll open an issue later
